src/shared_library.cpp

```diff
@@ -40,11 +40,32 @@ SharedLibrary::SharedLibrary(const std::string & library_path)
 }
 
 SharedLibrary::~SharedLibrary()
+{
+  if (lib.lib_pointer != NULL) {
+    rcutils_ret_t ret = rcutils_unload_shared_library(&lib);
+    if (ret != RCUTILS_RET_OK) {
+      std::cerr << rcutils_get_error_string().str << std::endl;
+      rcutils_reset_error();
+    }
+  } else {
+    if (lib.library_path != NULL) {
+      if (rcutils_allocator_is_valid(&lib.allocator)) {
+        lib.allocator.deallocate(lib.library_path, lib.allocator.state);
+        lib.allocator = rcutils_get_zero_initialized_allocator();
+      } else {
+        std::cerr << "ShareLibrary not able to deallocate memory on constructor" << std::endl;
+      }
+    }
+  }
```

Comment:
Why does this logic need to be here? Can't this call the unload API when a certain condition is met instead? Having knowledge about how the stuff needs to be deallocated etc. seems to break encapsulation.

Reply:
fixed https://github.com/ros2/rcpputils/pull/50/commits/cb11c673b72e18723647486e19bce1c06dcace72